history.php

```diff
@@ -21,17 +21,36 @@
  * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
 
+use tool_clonecategory\form\clonecategoryhistory_form;
 use tool_clonecategory\history_table;
 
 require('../../../config.php');
 require_once($CFG->libdir . '/adminlib.php');
 
 admin_externalpage_setup('clonecategory_history');
 
+$start = optional_param('start', strtotime('-3 month', time()), PARAM_INT);
+$end = optional_param('end', time(), PARAM_INT);
+
+$clonehistoryform = new clonecategoryhistory_form(null, [
+   'startdate' => $start,
+   'enddate' => $end,
+]);
+
+// Redirect back if form is cancelled.
+if ($clonehistoryform->is_cancelled()) {
+    redirect(new moodle_url("/admin/tool/clonecategory/history.php"));
+}
+
 $PAGE->set_url(new moodle_url("/admin/tool/clonecategory/history.php"));
```

Comment:
Would it be possible to `write_close` the session somewhere on this page ?

If the table takes 20s to output users will probably try switching to other tabs and they'll get a session lock error.

This should be possible since its just outputting data and not writing anything.

Reply:
Have added write_close in this last commit